src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+                }
+
+                Log($"Starting check for project '{project.FullPath}'.");
+
+                var timestampCache = _fileTimestampCache != null ? _fileTimestampCache.Value.TimestampCache : new Dictionary<string, DateTime>(StringComparer.OrdinalIgnoreCase);
+
+                DateTime GetLatestTimestamp(List<string> paths)
+                {
+                    var latestTime = DateTime.MinValue;
+
+                    foreach (var path in paths)
+                    {
+                        if (!timestampCache.TryGetValue(path, out var time))
+                        {
+                            if (File.Exists(path))
+                            {
+                                time = File.GetLastWriteTimeUtc(path);
+                                timestampCache[path] = time;
+                                Log($"Path '{path}' had live timestamp '{time}'.");
+                            }
+                            else
+                            {
+                                Log($"Path '{path}' did not exist.");
+                            }
+                        }
+                        else
+                        {
+                            Log($"Path '{path}' had cached timestamp '{time}'.");
+                        }
+
+                        if (latestTime < time)
+                        {
+                            latestTime = time;
+                        }
+                    }
+
+                    return latestTime;
+                }
+
+                if (_projectItemsSchema == null)
+                {
+                    Log($"Skipping because '{typeof(IProjectItemSchemaService).FullName}' component could not be found.");
+                    Report.IfNotPresent(_projectItemsSchema);
+                    return false;
+                }
+
+                if (!string.IsNullOrEmpty(project.GetPropertyValue(DisableFastUpToDateCheckProperty)))
+                {
+                    Log($"Disabled because the '{DisableFastUpToDateCheckProperty}' property is set to a non-empty value.");
+                    return false;
+                }
+
+                var inputs = new List<string>();
+
+                // add the project file
+                if (!string.IsNullOrEmpty(project.FullPath))
+                {
+                    inputs.Add(project.FullPath);
+                    Log($"Adding input project file {project.FullPath}.");
+                }
+
+                // add all imports except .user file
+                var userFilePath = project.FullPath + ".user";
+                var imports = project.Imports
+                    .Select(i => i.ImportedProject.FullPath)
+                    .Where(p => !string.IsNullOrEmpty(p) && !string.Equals(p, userFilePath, StringComparison.OrdinalIgnoreCase));
+
+                foreach (var import in imports)
+                {
+                    Log($"Adding input project import {import}.");
+                }
+
```

Comment:
this code actually run through the LinQ expression twice.  You should just do:
var imports = .....ToList(), and use the final list (and delete the inputs variable), instead of executing LinQ path twice.